### PR TITLE
Write server-side endpoints for custom repositories [SCI-2068]

### PR DESCRIPTION
**Pull request by ZmagoD**
https://biosistemika.atlassian.net/browse/SCI-2068
https://biosistemika.atlassian.net/browse/SCI-2070

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -449,7 +449,7 @@ def draw(routes_name)
 
     resources :repositories do
       post 'repository_index',
-           to: 'repositories#repository_table_index',
+           to: 'repository_rows#index', # repository_rows#index repositories#repository_table_index
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply config/routes.rb]: Can you please fix this?

File: app/models/repository_cell.rb [hunk 2]
```diff
@@ -8,6 +8,10 @@ class RepositoryCell < ActiveRecord::Base
   validates :repository_row,
             uniqueness: { scope: :repository_column }
 
+  belongs_to :repository_text_value, optional: true, foreign_key: :value_id
+  belongs_to :repository_date_value, optional: true, foreign_key: :value_id
+  belongs_to :repository_list_value, optional: true, foreign_key: :value_id
+
```
Comment (okriuchykhin) [inline app/models/repository_cell.rb]: This was updated with recent commits, so please don't overwrite it when merging

File: app/services/repository_datatable_service.rb [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+    order_by_index = ActiveRecord::Base.send(
+      :sanitize_sql_array,
+      ["position((',' || repository_rows.id || ',') in ?)",
+       ids.join(',') + ',']
+    )
+    records.order(order_by_index)
+  end
+
+  def select_type(type, records, id, dir)
```
Comment (okriuchykhin) [inline app/services/repository_datatable_service.rb]: Can we extend this from addon?

File: app/helpers/repository_datatable_helper.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            team: team
+          )
+      end
+      parsed_records << row
+    end
+    parsed_records
+  end
+
+  def assigned_row(record, assigned_rows)
+    if assigned_rows&.include?(record)
```
Comment (okriuchykhin) [inline app/helpers/repository_datatable_helper.rb]: assigned_rows& is it a typo?

File: app/models/repository_cell.rb (hunk 2 above)
Comment (ZmagoD) [reply app/models/repository_cell.rb]: :+1:

File: app/helpers/repository_datatable_helper.rb (hunk 4 above)
Comment (ZmagoD) [reply app/helpers/repository_datatable_helper.rb]: assigned_rows&.include?(record)  from ruby 2.2 is shortcut for  assigned_rows && assigned_rows.include?...

File: app/services/repository_datatable_service.rb [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                                 .preload(
+                                   :repository_columns,
+                                   :created_by,
+                                   repository_cells: :value
+                                 )
+                                 .joins(:created_by)
+                                 .where(repository: @repository)
+      return @assigned_rows if @params[:assigned] == 'assigned'
+    else
+      @assigned_rows = repository_rows.joins(
```
Comment (okriuchykhin) [inline app/services/repository_datatable_service.rb]: I think we can use here simply '@assigned_rows = repository_rows.joins(:my_module_repository_rows)'

File: app/services/repository_datatable_service.rb [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      dir == column_obj[:dir].upcase
+    end || 'ASC'
+    column_index = column_obj[:column]
+    col_order = @repository.repository_table_states
+                           .find_by_user_id(@user.id)
+                           .state['ColReorder']
+    column_id = col_order[column_index].to_i
+
+    if sortable_columns[column_id - 1] == 'assigned'
+      if @my_module && @params[:assigned] == 'assigned'
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/repository_datatable_service.rb]: Can you fix this?

**Pull request merged**